File: cpp/77. Combinations.cpp
#include <bits/stdc++.h>
using namespace std;
typedef long long ll;
typedef vector<int> vi;
typedef vector<vector<int>> vvi;
typedef pair<int, int> pi;

class Solution {
public:
    vvi ans;
    void dfs(int i, vi &path, int n, int k) {
        if(path.size() == k) {
            ans.push_back(path);
            return;
        }
        for(int j = i; j <= n; j++) {
            path.push_back(j);
            dfs(j+1, path, n, k);
            path.pop_back();
        }
    }
    vector<vector<int>> combine(int n, int k) {
        vi path;
        dfs(1, path, n, k);
        return ans;
    }
};

void printAns(vvi ans) {
    for(auto l: ans) {
        cout << "(";
        for(auto n: l) {
            cout << n << " ";
        }
        cout << "), ";
    }
    cout << "\n";
};

int main() {
    ios::sync_with_stdio(0); 
    cin.tie(0);

    Solution solution; int n, k; vvi ans;

    solution = Solution();
    n = 4, k = 2;
    ans = solution.combine(n, k);
    printAns(ans);

    solution = Solution();
    n = 1, k = 1;
    ans = solution.combine(n, k);
    printAns(ans);
}

File: python/77-Combinations.py
from typing import List


class Solution:
    # def combine(self, n: int, k: int) -> List[List[int]]:
    #     "Time: O(n^k), Space: O(k)"
    #     result = []
    #     def dfs(i, path):
    #         path.append(i)
    #         if len(path) == k:
    #             result.append(path)
    #             return
    #         for j in range(i+1, n+1):
    #             dfs(j, path[:])
    #     for i in range(1, n+1):
    #         dfs(i, [])
    #     return result
    def combine(self, n: int, k: int) -> List[List[int]]:
        "Time: O(n^k), Space: O(k)"
        result = []
        def dfs(i, path):
            if len(path) == k:
                result.append(path)
                return
            for j in range(i+1, n+1):
                dfs(j, path+[j])
        dfs(0, [])
        return result

n = 4
k = 2
result = Solution().combine(n, k)
print(result) # [[2,4],[3,4],[2,3],[1,2],[1,3],[1,4],]

n = 1
k = 1
result = Solution().combine(n, k)
print(result) # [[1]]

